File: android-app/src/main/java/fr/cph/chicago/core/fragment/SettingsFragment.java
package fr.cph.chicago.core.fragment;

import android.app.AlarmManager;
import android.app.AlertDialog;
import android.app.PendingIntent;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.io.File;
import java.util.Random;

import butterknife.BindView;
import fr.cph.chicago.R;
import fr.cph.chicago.core.App;
import fr.cph.chicago.core.activity.BaseActivity;
import fr.cph.chicago.repository.RealmConfig;
import fr.cph.chicago.service.PreferenceService;
import fr.cph.chicago.util.Util;

@SuppressWarnings("WeakerAccess")
public class SettingsFragment extends AbstractFragment {

    @BindView(R.id.clear_cache)
    LinearLayout clearCache;

    @BindView(R.id.version_number)
    TextView versionNumber;

    private final Util util;
    private final PreferenceService preferenceService;
    private final RealmConfig realmConfig;

    public SettingsFragment() {
        util = Util.INSTANCE;
        preferenceService = PreferenceService.INSTANCE;
        realmConfig = RealmConfig.INSTANCE;
    }

    @NonNull
    public static SettingsFragment newInstance(final int sectionNumber) {
        return (SettingsFragment) fragmentWithBundle(new SettingsFragment(), sectionNumber);
    }

    @Override
    public final void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        util.trackScreen((App) getActivity().getApplication(), getString(R.string.analytics_settings_fragment));
    }

    @Override
    public final View onCreateView(final LayoutInflater inflater, final ViewGroup container, final Bundle savedInstanceState) {
        final View rootView = inflater.inflate(R.layout.fragment_settings, container, false);
        if (!activity.isFinishing()) {
            setBinder(rootView);
            final String version = "Version " + util.getCurrentVersion(getContext());
            versionNumber.setText(version);
            clearCache.setOnClickListener(view -> {
                final DialogInterface.OnClickListener dialogClickListener = (dialog, which) -> {
                    switch (which) {
                        case DialogInterface.BUTTON_POSITIVE:
                            cleanLocalData();
                            restartApp();
                            break;
                        case DialogInterface.BUTTON_NEGATIVE:
                            break;
                    }
                };

                new AlertDialog.Builder(getContext())
                    .setMessage("This is going to:\n\n- Delete all your favorites\n- Clear application cache\n- Restart the application")
                    .setPositiveButton("Yes", dialogClickListener)
                    .setNegativeButton("No", dialogClickListener)
                    .show();
            });
        }
        return rootView;
    }

    private void restartApp() {
        final Intent intent = new Intent(getContext(), BaseActivity.class);
        final int intentId = new Random().nextInt();
        final PendingIntent pendingIntent = PendingIntent.getActivity(getContext(), intentId, intent, PendingIntent.FLAG_CANCEL_CURRENT);
        final AlarmManager alarmManager = (AlarmManager) getContext().getSystemService(Context.ALARM_SERVICE);
        alarmManager.set(AlarmManager.RTC, System.currentTimeMillis(), pendingIntent);
        activity.finish();
    }

    private void cleanLocalData() {
        deleteCache(getContext());
        preferenceService.clearPreferences(getContext());
        realmConfig.cleanRealm(getContext());
    }

    private void deleteCache(final Context context) {
        try {
            final File cacheDirectory = context.getCacheDir();
            deleteRecursiveDirectory(cacheDirectory);
        } catch (Exception ignored) {
        }
    }

    private boolean deleteRecursiveDirectory(final File directory) {
        if (directory != null && directory.isDirectory()) {
            final String[] children = directory.list();
            for (final String child : children) {
                boolean success = deleteRecursiveDirectory(new File(directory, child));
                if (!success) {
                    return false;
                }
            }
            return directory.delete();
        } else
            return directory != null && directory.isFile() && directory.delete();
    }
}

File: android-app/src/main/kotlin/fr/cph/chicago/core/fragment/SettingsFragment.kt
/**
 * Copyright 2021 Carl-Philipp Harmant
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.cph.chicago.core.fragment

import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AlertDialog
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import fr.cph.chicago.Constants.SELECTED_ID
import fr.cph.chicago.R
import fr.cph.chicago.core.App
import fr.cph.chicago.core.activity.BaseActivity
import fr.cph.chicago.core.activity.DeveloperOptionsActivity
import fr.cph.chicago.core.model.Theme
import fr.cph.chicago.databinding.FragmentSettingsBinding
import fr.cph.chicago.redux.ResetStateAction
import fr.cph.chicago.redux.store
import fr.cph.chicago.repository.RealmConfig
import fr.cph.chicago.service.PreferenceService

class SettingsFragment : Fragment(R.layout.fragment_settings) {

    companion object {
        private val preferenceService = PreferenceService
        private val realmConfig = RealmConfig

        fun newInstance(sectionNumber: Int): SettingsFragment {
            return fragmentWithBundle(SettingsFragment(), sectionNumber) as SettingsFragment
        }
    }

    private var _binding: FragmentSettingsBinding? = null
    private val binding get() = _binding!!

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        _binding = FragmentSettingsBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        val version = "Version ${util.getCurrentVersion()}"
        binding.versionNumber.text = version
        binding.themeName.text = preferenceService.getTheme().description

        binding.theme.setOnClickListener {
            val builder = MaterialAlertDialogBuilder(context!!, R.style.AlertDialog)
            val choices = Theme.values().map { it.description }.toTypedArray()
            val selected = choices.indexOf(preferenceService.getTheme().description)
            builder.setTitle("Theme change")
            builder.setSingleChoiceItems(choices, selected, null)
            builder.setPositiveButton("Save") { dialog: DialogInterface, _ ->
                val list = (dialog as AlertDialog).listView
                for (i in 0 until list.count) {
                    val checked = list.isItemChecked(i)
                    if (checked) {
                        preferenceService.saveTheme(Theme.values()[i])
                        reloadActivity()
                    }
                }
            }
            builder.setNegativeButton("Cancel", null)
            builder.show()
        }

        binding.clearCache.setOnClickListener {
            val dialogClickListener = { _: Any, which: Any ->
                when (which) {
                    DialogInterface.BUTTON_POSITIVE -> {
                        cleanLocalData()
                        restartApp()
                    }
                    DialogInterface.BUTTON_NEGATIVE -> {
                    }
                }
            }

            MaterialAlertDialogBuilder(context!!, R.style.AlertDialog)
                .setMessage("This is going to:\n\n- Delete all your favorites\n- Clear application cache\n- Reload the application")
                .setPositiveButton("Yes", dialogClickListener)
                .setNegativeButton("No", dialogClickListener)
                .show()
        }

        binding.developerLayout.setOnClickListener {
            // start new developer activity
            val intent = Intent(context, DeveloperOptionsActivity::class.java)
            context?.startActivity(intent)
        }
    }

    private fun reloadActivity() {
        App.instance.themeSetup()
        val intent = activity?.intent
        intent?.putExtra(SELECTED_ID, R.id.navigation_settings)
        intent?.putExtra(getString(R.string.bundle_title), getString(R.string.settings))
        activity?.finish()
        startActivity(activity?.intent)
    }

    private fun restartApp() {
        store.dispatch(ResetStateAction())
        val intent = Intent(context, BaseActivity::class.java)
        activity?.finish()
        startActivity(intent)
    }

    private fun cleanLocalData() {
        deleteCache(context)
        preferenceService.clearPreferences()
        realmConfig.cleanRealm()
    }

    private fun deleteCache(context: Context?) {
        try {
            context?.cacheDir?.deleteRecursively()
        } catch (ignored: Exception) {
        }
    }
}
